evm/vm/forks/sharding/contracts/validator_manager.v.py

```diff
@@ -242,115 +191,65 @@ def get_eligible_proposer(shard_id: num, period: num) -> address:
-                            )
-                    ),
-                    as_num256(validators_max_index)
-                )
-                if valcode_addr == self.validators[as_num128(validator_index)].validation_code_addr:
-                    shard_list[shard_id] = True
-                    break
-    return shard_list
-
-
-# emit a log which is equivalent to
-# CollationAdded: __log__({shard_id: indexed(num), collation_header: bytes <= 4096, is_new_head: bool, score: num})
-# TODO: should be replaced by `log.CollationAdded`
-@internal
-def emit_collation_added(shard_id: num, collation_header: bytes <= 4096, is_new_head: bool, score: num):
-    if is_new_head:
-        new_head_in_num = 1
-    else:
-        new_head_in_num = 0
-    raw_log(
-        [
-            sha3("CollationAdded(int128,bytes4096,bool,int128)"),
-            as_bytes32(shard_id),
-        ],
-        concat(
-            collation_header,
-            as_bytes32(new_head_in_num),
-            as_bytes32(score),
-        )
-    )
+    ].addr
 
 
 # Attempts to process a collation header, returns True on success, reverts on failure.
 @public
-def add_header(header: bytes <= 4096) -> bool:
+def add_header(
+        shard_id: num,
+        expected_period_number: num,
+        period_start_prevhash: bytes32,
+        parent_hash: bytes32,
+        transaction_root: bytes32,
+        collation_coinbase: address,  # TODO: cannot be named `coinbase` since it is reserved
+        state_root: bytes32,
+        receipt_root: bytes32,
+        collation_number: num,  # TODO: cannot be named `number` since it is reserved
+    ) -> bool:
     zero_addr = 0x0000000000000000000000000000000000000000
 
-    values = RLPList(header, [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes])
-    shard_id = values[0]
-    expected_period_number = values[1]
-    period_start_prevhash = values[2]
-    parent_collation_hash = values[3]
-    tx_list_root = values[4]
-    collation_coinbase = values[5]
-    post_state_root = values[6]
-    receipt_root = values[7]
-    collation_number = values[8]
-    sig = values[9]
-
     # Check if the header is valid
     assert (shard_id >= 0) and (shard_id < self.shard_count)
     assert block.number >= self.period_length
     assert expected_period_number == floor(decimal(block.number / self.period_length))
     assert period_start_prevhash == blockhash(expected_period_number * self.period_length - 1)
 
     # Check if this header already exists
-    entire_header_hash = sha3(header)
+    header_bytes = concat(
+        as_bytes32(shard_id),
+        as_bytes32(expected_period_number),
+        period_start_prevhash,
+        parent_hash,
+        transaction_root,
+        as_bytes32(collation_coinbase),
+        state_root,
+        receipt_root,
+        as_bytes32(collation_number),
+    )
+    entire_header_hash = sha3(header_bytes)
     assert entire_header_hash != as_bytes32(0)
```

Comment:
Unnecessary assert

Reply:
You're right. Thank you for pointing out.